src/orion/algo/space/__init__.py

```diff
@@ -67,6 +71,64 @@ def __repr__(self):
         return "..."
 
 
+def _to_snake_case(name: str) -> str:
+    """Transform a class name ``MyClassName`` to snakecase ``my_class_name``"""
+    frags = []
+
+    frag = []
+    for c in name:
+        if c.isupper() and frag:
+            frags.append("".join(frag).lower())
+            frag = []
+
+        frag.append(c)
+
+    if frag:
+        frags.append("".join(frag).lower())
+
+    return "_".join(frags)
+
+
+T = TypeVar("T")
+
+
+class SpaceConverter(Generic[T]):
+    """SpaceConverter iterates over an Orion search space.
+    This can be used to implement new features for ``orion.algo.space.Space``
+    outside of Orion's code base.
+
+    """
+
+    def convert_dimension(self, dimension: Dimension) -> T:
+        """Call the dimension conversion handler"""
+        return getattr(self, _to_snake_case(type(dimension).__name__))(dimension)
```

Comment:
This is essentially a singledispatch callable in disguise: You have handlers for each type below, and you dispatch based on the "type" of the argument, indirectly.
However, having an implicit dependency on the name of the dimension classes is very brittle, and hard to keep track of. Could you make this more explicit somehow?
I also don't think this should be inside a class, since there isn't (and most likely won't be) any state.

Reply:
Also, what happens if a new type of dimension is added, and there is no method with that name? This will raise an `AttributeError`, but you'd ideally like a `NotImplementedError` with an informative message.

Reply:
If we make it an abstract class then non-implemented methods would already raise an informative message during code development. It's true there is no state so making it a class isn't super useful for functionality's sake but I think it is nevertheless useful to make the interface more clear. The other solution would be to have a module for these functions only (without the methods for the conversion the other way around).

In any case, we should make it such that if the dev forgets to write some methods it fails with a clear message at execution or some tests automatically detects this.

Reply:
Actually, this should use the property `type` to work with transformed dimensions. Looks like we should also have tests for converting transformed spaces. 
```suggestion
    def convert_dimension(self, dimension: Dimension) -> T:
        """Call the dimension conversion handler"""
        return getattr(self, _to_snake_case(dimension.type))(dimension)
```

Reply:
As long as:
- we don't dispatch based on the names of classes, and 
- the solution is consistent with itself and the base class,
then I don't have a problem with using an ABC and marking all the methods with @abstractmethod.

Reply:
> It's true there is no state so making it a class isn't super useful for functionality's sake but I think it is nevertheless useful to make the interface more clear. The other solution would be to have a module for these functions only (without the methods for the conversion the other way around).

Since the abstract base class would only have static methods, this is actually a good use-case for [Protocols](https://docs.python.org/3/library/typing.html#typing.Protocol):
```python
from typing_extensions import Protocol

T = TypeVar("T")

class SpaceConverter(Protocol[T]):
     integer: Callable[[Integer], T]
     real: Callable[[Real], T]
     fidelity: Callable[[Fidelity], T | None]
```

This way, if you have a module like so:
```python
# nevergrad/conversion.py (or something like that)

def integer(dim: Integer) -> NeverGradStuff:
     ...
    
def real(dim: Real) -> NeverGradReal:
    ...
```

The procol can be used just like the abstract base class, and has the benefit of allowing type checkers to also check against a module:
```python
from nevergrad import conversion

space_converter: SpaceConverter[NeverGradStuff] = conversion    # Passes, since all the required methods are defined.
 ```